ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/AggregateImpressions.java

```diff
@@ -0,0 +1,109 @@
+package com.mozilla.telemetry.contextualservices;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.mozilla.telemetry.ingestion.core.Constant.Attribute;
+import com.mozilla.telemetry.transforms.PubsubConstraints;
+import com.mozilla.telemetry.util.Json;
+import com.mozilla.telemetry.util.Time;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessageWithAttributesCoder;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.WithKeys;
+import org.apache.beam.sdk.transforms.WithTimestamps;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.IntervalWindow;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Instant;
+
+/**
+ * Count the number of impressions aggregated by reporting URL.
+ */
+public class AggregateImpressions
+    extends PTransform<PCollection<PubsubMessage>, PCollection<PubsubMessage>> {
+
+  private final List<String> aggregationFields = ImmutableList.of(
+      ReportingUrlUtil.PARAM_COUNTRY_CODE, ReportingUrlUtil.PARAM_REGION_CODE,
+      ReportingUrlUtil.PARAM_FORM_FACTOR, ReportingUrlUtil.PARAM_OS_FAMILY,
+      ReportingUrlUtil.PARAM_ID);
+
+  private final ValueProvider<String> getAggregationWindowDuration;
+
+  public AggregateImpressions(ValueProvider<String> getAggregationWindowDuration) {
+    this.getAggregationWindowDuration = getAggregationWindowDuration;
+  }
+
+  public static AggregateImpressions of(ValueProvider<String> getAggregationWindowSize) {
+    return new AggregateImpressions(getAggregationWindowSize);
+  }
+
+  @Override
+  public PCollection<PubsubMessage> expand(PCollection<PubsubMessage> messages) {
+    return messages //
+        .apply(WithKeys.of((SerializableFunction<PubsubMessage, String>) this::getAggregationKey)) //
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), PubsubMessageWithAttributesCoder.of())) //
+        .apply(WithTimestamps.of(message -> new Instant())) //
+        .apply(Window.into(FixedWindows.of(Time.parseDuration(getAggregationWindowDuration.get())))) //
```

Comment:
I'm realizing that we may not want to set a window at all, and instead use GlobalWindows with a processing time trigger: https://beam.apache.org/documentation/programming-guide/#processing-time-triggers

It would be something like:

```
        .apply(Window.configure().triggering(
          AfterProcessingTime.pastFirstElementInPane().alignedTo(
            Duration.standardMinutes(10))))
```

I think that eliminates the possibility of dropped messages, as this would still drain remaining messages when the pipeline is drained.

Problems with this: it appears that `AfterProcessingTime` doesn't accept ValueProvider. Also, I haven't fully been able to figure out syntax here to get global windows with this trigger.

It would be interesting to at least try out this approach and see if we see the dropped message problem go away.

Reply:
After some reading and tinkering this is what I came up with before my EOD:
```java
return messages //
    .apply(WithKeys.of((SerializableFunction<PubsubMessage, String>) this::getAggregationKey)) //
    .setCoder(KvCoder.of(StringUtf8Coder.of(), PubsubMessageWithAttributesCoder.of())) //
    .apply(Window.<KV<String, PubsubMessage>>configure()
        .triggering(Repeatedly.forever(AfterProcessingTime.pastFirstElementInPane()
            .alignedTo(Time.parseDuration(aggregationWindowDuration.get()))))
        .discardingFiredPanes()) //
    .apply(Count.perKey()) //
    .apply(ParDo.of(new BuildAggregateUrl())) //
    .apply(Window.<PubsubMessage>into(new GlobalWindows()).triggering(DefaultTrigger.of()));
```

Setting default trigger after is needed because of `An exception occured while executing the Java class. Inputs to Flatten had incompatible triggers: Repeatedly.forever(AfterSynchronizedProcessingTime.pastFirstElementInPane()), DefaultTrigger` from the write output stage.  I'm not sure if this is right thing to do.

This uses a global window instead of an interval window which means we can't get a nice start and end of the aggregation period without some extra steps.

This seems to do the right thing.  Urls with aggregated counts are being output but I haven't validated it.  Also I don't see any droppedDueToLateness after stopping the job.